V2/src/fc_state_machine.cpp

```diff
@@ -2,57 +2,164 @@
 
 #include <fc_state_machine.h>
 #include <error_checker.h>
-
 #include <pin_defs.h>
+#include <analogs.h>
+
+#define CAP_THRESHOLD 20 // Move to another file, if they aren't already.
+
+void update_leds(void);
 
 #ifdef ENABLE_FC_TEST
 uint32_t state = FC_TEST;
 #else
-uint32_t state = FC_INIT;
+uint32_t state = FC_STANDBY;
 #endif
 
-uint32_t get_fc_state() {
-    // read-only access to state variable
+DigitalIn start(START); 
+
+// Valves.
+DigitalOut supply_v(SUPPLY_V);
+DigitalOut purge_v(PURGE_V);
+DigitalOut other1_v(VALVE3);
+DigitalOut other2_v(VALVE4);
+
+// Relays.
+DigitalOut start_r(START_R);
+DigitalOut motor_r(MOTOR_R);
+DigitalOut charge_r(CHARGE_R);
+DigitalOut cap_r(CAP_R);
+DigitalOut fcc_r(FCC_R);
+
+// LEDs.
+DigitalOut alarm_led(ALARM_LED);
+DigitalOut debug_led(DEBUG_LED);
+DigitalOut shut_led(SHUT_LED);
+DigitalOut run_led(RUN_LED);
+DigitalOut start_led(START_LED);
+
+uint32_t get_fc_state() // Read-only access to state variable.
+{
     return state;
 }
 
-// todo - other digital pin defs
-DigitalIn start(START);
+void fc_state_machine_thread() 
+{
+    fcc_r.write(true); // Solid state relay.
 
-// todo - define which timers we need
-Timer timer;
-
-void fc_state_machine_thread() {
-    while (true) {
-        if (check_all_errors()) {
-            state = FC_FAULT;
-        }
+    while (true) 
+    {
+        if (check_all_errors()) 
+            state = FC_ALARM;
 
-        if (state == FC_TEST) {
+        if (FC_TEST == state) 
+        {
             ThisThread::sleep_for(50); // Add test state code here, if desired.
+            charge_r.write(false);
         }
 
-        if (state == FC_INIT){
-            if (start) {
-                state = FC_STARTUP;
-                // we will need lots if timers
-                // timer example:
-                timer.start();
+        if (FC_STANDBY == state) // Wait until start button is pressed.
+        { 
+            if (start) 
+            {
+                purge_v.write(false); // Perform start purge.
+                start_r.write(true);
+                motor_r.write(false);
+                charge_r.write(false);
+                cap_r.write(false);
+                supply_v.write(true);
+                Thread::wait(1000);
+                purge_v.write(true);
+                Thread::wait(200);
+                purge_v.write(false); 
+
+                supply_v.write(true); // Charge state setup.
+                start_r.write(false);
+                motor_r.write(false);
+                charge_r.write(true);
+                cap_r.write(false);
```

Comment:
I think it would be good to add a h2 pressure check before connecting a load so we don't accidentally fuel starve the stack (If someone presses the start button before turning on h2). 
Also put the startup purge as a separate state
Also I think according to ballard we are supposed to connect a load during the startup purge (double check this though)
Example:
```
if state == STANDBY
    if start
        open supply value
        state = PRESSURIZE
if state == PRESSURIZE
    sleep 1000 // might not need this
    if PRES > THRES
        state = STARTUP_PURGE
if state == STARTUP_PURGE
    open purge valve
    close startup relay
    sleep 200 // there is a chance this is supposed to be the 1 sec above
    close purge valve
    open startup relay
    state = charge
    put charge state setup here
if state = CHARGE
   etc
```

Reply:
We are actually supposed to connect to the startup resistor right before the startup purge (so close the startup relay before we open the purge valve). The startup resistor is connected across the resistor terminals on the relay board.

Reply:
Will add purge state on next commit. Is the connection to the startup resistor supposed to happen differently than I have it? Startup relay is closed (65), startup purge happens, startup relay opens (76).

Reply:
Also I think we should do the startup purge as soon as the pressure reaches a threshold OR after a certain time interval. Just in case something is wrong and we don't reach the threshold, I think it is still important to purge.

Reply:
Ya my only concern with that is if there is no hydrogen connected then you probably don't want to apply a load. Can set the threshold really low

Reply:
> 
> 
> Will add purge state on next commit. Is the connection to the startup resistor supposed to happen differently than I have it? Startup relay is closed (65), startup purge happens, startup relay opens (76).

I didn't see it there. probably fine